src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -432,9 +419,9 @@ bool LateInlineMHCallGenerator::do_late_inline_check(Compile* C, JVMState* jvms)
   assert(!input_not_const, "sanity"); // shouldn't have been scheduled for inlining in the first place
 
   if (cg != nullptr) {
-    if (!allow_inline && (C->print_inlining() || C->print_intrinsics())) {
-      C->print_inlining(cg->method(), jvms->depth()-1, call_node()->jvms()->bci(), InliningResult::FAILURE,
-                        "late method handle call resolution");
+    if (!allow_inline) {
```

Comment:
Is `!allow_inline` really the correct check here and in LateInlineVirtualCallGenerator::do_late_inline_check()?  This is where I ran into trouble with the asserts in previous implementation.  Note that even if allow_inline is passed as true to the CallGenerator factory, the factory can force it to false for a variety of reasons.  Maybe we should be looking at what kind of CallGenerator we have in `cg`?

Reply:
With the new implementation it is always safe to call C->inline_printer()->record. In case inline printing is disabled, it is just a no-op. Also allow_inline seems to be coming from C->inlining_incrementally(). Is your concern that we might miss to print something? Or that we print something extra that is not true?

Reply:
> With the new implementation it is always safe to call C->inline_printer()->record. In case inline printing is disabled, it is just a no-op. Also allow_inline seems to be coming from C->inlining_incrementally(). Is your concern that we might miss to print something?

Yes, that we might miss printing something when allow_inline was true.

Reply:
I think I didn't change the logic behind when it would print now as compared to before my refactoring. I think investigating this is a bit beyond the scope about this PR. Should I file a RFE for this?

Reply:
Right, it's an existing issue not related to your changes.  Yes, please file an RFE.